### PR TITLE
IDEMPIERE-4622  Cross Tenant UserDefInfo, Role

### PR DESCRIPTION
**Pull request by igorpojzl**
https://idempiere.atlassian.net/browse/IDEMPIERE-4622

### CODE REVIEW

File: org.adempiere.base/src/org/compiere/model/MUserDefInfo.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ public MUserDefInfo(Properties ctx, ResultSet rs, String trxName) {
 	private static MUserDefInfo[] getAll (Properties ctx, int infowindow_ID )
 	{
 		if (m_fullList == null) {
-			m_fullList = new Query(ctx, MUserDefInfo.Table_Name, "IsActive='Y'", null).list();
+			m_fullList = new Query(ctx, MUserDefInfo.Table_Name, "IsActive='Y'", null).setClient_ID().list();
```
Comment (CarlosRuiz-globalqss) [inline org.adempiere.base/src/org/compiere/model/MUserDefInfo.java:84]: This is wrong, as is designed m_fullList is intended to have a list from all tenants.

To make this work would require a refactor similar to what @hengsin did in MUserDefWin here:
https://github.com/idempiere/idempiere/commit/2ab2211f81f4ecb3928a56506ddd7712731a1036#diff-18f1e200c9da97cb1ba9cb7290831a270c9dd9d5a9721ff68628a16dd1b574fd

File: org.adempiere.base/src/org/compiere/model/MUserDefProc.java [hunk 2]
```diff
@@ -87,7 +87,7 @@ public MUserDefProc(Properties ctx, MUserDefProc copy, String trxName) {
 	private static MUserDefProc[] getAll (Properties ctx, int processID)
 	{
 		if (m_fullList == null) {
-			m_fullList = new Query(ctx, MUserDefProc.Table_Name, "IsActive='Y'", null).list();
+			m_fullList = new Query(ctx, MUserDefProc.Table_Name, "IsActive='Y'", null).setClient_ID().list();
```
Comment (CarlosRuiz-globalqss) [inline org.adempiere.base/src/org/compiere/model/MUserDefProc.java:90]: Same as comment above for MUserDefProc

File: org.adempiere.base/src/org/compiere/model/MUserRoles.java [hunk 3]
```diff
@@ -64,6 +64,7 @@
 		final String whereClause = I_AD_User_Roles.COLUMNNAME_AD_User_ID+"=?";	
 		List<MUserRoles> list = new Query(ctx,I_AD_User_Roles.Table_Name,whereClause,null)
 		.setParameters(AD_User_ID)
+		.setClient_ID()
```
Comment (CarlosRuiz-globalqss) [inline org.adempiere.base/src/org/compiere/model/MUserRoles.java:67]: This seems wrong, filtering by AD_Client_ID this way will exclude the roles the user can have in System tenant.

I think the filter must be something like adding in the whereClause
`AND AD_Client_ID IN (0,?)`
and adding the clientID to the parameters

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]

**Comment by hengsin** [comment]
For MUserDefInfo.m_fullList, please refer to the same changes that I've made for MUserDefWin (replace by m_fullMap).

**Comment by igorpojzl** [comment]
Commited all Requests.

**Comment by norbertbede** [comment]
@CarlosRuiz-globalqss @hengsin  , can you please help you sort out, if any investigation needed, we are available. asap.

File: org.adempiere.base/src/org/compiere/model/MUserDefProc.java [hunk 4]
```diff
@@ -87,7 +87,7 @@ public MUserDefProc(Properties ctx, MUserDefProc copy, String trxName) {
 	private static MUserDefProc[] getAll (Properties ctx, int processID)
 	{
 		if (m_fullList == null) {
-			m_fullList = new Query(ctx, MUserDefProc.Table_Name, "IsActive='Y'", null).list();
+			m_fullList = new Query(ctx, MUserDefProc.Table_Name, "IsActive='Y' AND AD_Client_ID IN (0,?)", null).setParameters(Env.getAD_Client_ID(ctx)).list();
```
Comment (CarlosRuiz-globalqss) [inline org.adempiere.base/src/org/compiere/model/MUserDefProc.java:90]: Hi @igorpojzl 

You must apply here the same approach using m_fullMap as in MUserDefInfo

Regards,

Carlos Ruiz

**Comment by igorpojzl** [comment]
Commited Changes.

**Review by CarlosRuiz-globalqss — APPROVED** [review]
Tested successfully

**Pull request merged**